File: Data Structures/Queue/C/queue.c
#include "queue.h"

/*
    This method handles the create of the Queue
*/
struct Queue *createQueue() {
    // allocate memory for the queue
    struct Queue *queue = malloc(sizeof(struct Queue));

    // if queue is null
    if (!queue) {
        // output error to user
        fprintf(stderr, "There was a problem initializing the Queue\n");
        // return null
        return NULL;
    }
    // set the capacity
    queue->capacity = SIZE;
    // init the front/rear to empty
    queue->front = queue->rear = EMPTY;
    // allocate memory for the array
    queue->data = malloc(queue->capacity * sizeof(int));
    // if the array is null
    if (!queue->data) {
        // output error to user
        fprintf(stderr, "There was a problem initializing the array for the Queue\n");
        // return null
        return NULL;
    }
    // return the queue
    return queue;
}

/* This method returns 1 if empty, 0 if not */
int isEmpty(struct Queue *Q) {
    return (Q->front == EMPTY);
}

/* This method returns 1 if full, 0 if not */
int isFull(struct Queue *Q) {
    // if the rear mod capacity is the same as the front
    // then this queue is full
    return ((Q->rear+1) % Q->capacity == Q->front);
}

/* This method returns the size of the queue */
int size(struct Queue *Q) {
    return (Q->capacity - Q->front + Q->rear + 1) % Q->capacity;
}

/* This method adds data to the queue */
void enQueue(struct Queue *Q, int data) {
    // if this queue is full
    if (isFull(Q))
        // resize by 2
        resize(Q, 2.0f);
    // change the rear based on the end mod capcity
    Q->rear = (Q->rear+1) % Q->capacity;
    // set the data to the rear of the queue
    Q->data[Q->rear] = data;

    // if front is -1
    if (Q->front == -1)
        // front and rear are the same point
        Q->front = Q->rear;
}

/* This method handles removing the first
    item in the queue */
int deQueue(struct Queue *Q) {
    // init data to 0
    int data = 0;
    // if this queue is empty
    if (isEmpty(Q))
        // return the data
        return data;
    else {
        // check if the size is less than half of capacity
        if (size(Q) < Q->capacity / 2)
            // resize by half
            resize(Q, 0.5f);
        // grab the data from the front
        data = Q->data[Q->front];
        // if one item in queue
        if (Q->front == Q->rear)
            // set front to rear - 1 
            Q->front = Q->rear - 1;
        else
            // set the front
            Q->front = (Q->front+1) % Q->capacity;
    }
    // return the data
    return data;
}

/* This method handles resizing the queue */
void resize(struct Queue *Q, float factor) {
    // get the capacity
    int size = Q->capacity;
    // change the capacity by the factor
    Q->capacity = (int) (size * factor);
    // reallocate the memory
    Q->data = realloc(Q->data, Q->capacity);
    // if the array is null
    if (!Q->data) {
        // output error message to user
        fprintf(stderr, "There was an error reallocating memory\n");
        // and return
        return;
    }
    // as long as the front is before the rear of the queue
    if (Q->front > Q->rear) {
        // loop over the array to front
        int i;
        for (i=0; i<Q->front; i++)
            // copy over
            Q->data[i+size] = Q->data[i];
        // set th rear
        Q->rear = Q->rear + size;
    }
}

File: Data Structures/Queue/C#/Queue.cs
using System;

namespace DataStructures.Classes
{
    /// <summary>
    /// This class is the Queue ADT
    /// </summary>
    /// <typeparam name="T">Generic Type</typeparam>
    public class Queue<T>
    {
        // the default capcity of the queue
        private const int SIZE = 16;

        
        private int front;    // location of the front of the queue
        private int back;     // location of the back of the queue
        private int size;     // the capacity of the queue
        private int N;        // the number of items in the queue
        private T[] queue;    // the queue itself, an array
        
        /// <summary>
        /// Default constructor
        /// </summary>
        public Queue()
        {
            Init(SIZE);
        }

        /// <summary>
        /// Constructor with optional capacity parameter
        /// </summary>
        /// <param name="size">the capacity of the queue</param>
        public Queue(int size)
        {
            Init(size);
        }

        /// <summary>
        /// Private helper method to init the object
        /// upon constructor being called
        /// </summary>
        /// <param name="size">size of the queue</param>
        private void Init(int size)
        {
            // set the size
            this.size = size;
            // init the queue array with size
            queue = new T[this.size];
            // set front and back to -1
            front = back = -1;
            // start the number of items at 0
            N = 0;
        }

        /// <summary>
        /// Gets the number of items currently in the queue
        /// </summary>
        /// <returns>The number of items in queue</returns>
        public int Length() => N;

        /// <summary>
        /// Gets the capcacity of the queue
        /// </summary>
        public int Capacity => size;

        /// <summary>
        /// Returns true if queue is empty
        /// </summary>
        public bool IsEmpty => N == 0;
        
        /// <summary>
        /// Puts in the back of the queue the item
        /// </summary>
        /// <param name="data">The item to enqueue</param>
        public void EnQueue(T data)
        {
            // if empty, set the front and back to 0
            if (IsEmpty) front = back = 0;
            else
            {
                // set back to the next in queue
                // note that this will wrap around the array
                // if user goes over the capacity, then will write over
                // TODO: either throw exception or resize array
                back = (back + 1) % size;
            }
            // set the data in the back of the queue
            queue[back] = data;
            // increment the number of items in the queue
            N++;
        }

        /// <summary>
        /// Removes the front item from the queue
        /// </summary>
        /// <returns>The front item</returns>
        /// <exception cref="ArgumentOutOfRangeException">Throws exception if queue is empty</exception>
        public T DeQueue()
        {
            // if empty, throw exception
            if (IsEmpty)
                throw new ArgumentOutOfRangeException($"Queue is empty");

            // cache the item from the front of the queue
            T item = queue[front];

            // since we are using the ability to wrap around the queue
            // then once front and back are the same, then queue is
            // now empty, so reset back to -1
            if (front == back)
                front = back = -1;
            else
                // set front to the next
                front = (front + 1) % size;
            // decrement to number of items in the queue
            N--;
            // return the item
            return item;
        }
    }
}